include/chainbase/undo_index.hpp

```diff
@@ -0,0 +1,800 @@
+         void undo() {
+            if ( _apply ) _index.undo();
+            _apply = false;
+         }
+       private:
+         undo_index& _index;
+         bool _apply = true;
+      };
+
+      int64_t revision() const { return _revision; }
+
+      session start_undo_session( bool enabled ) {
+         return session{*this, enabled};
+      }
+
+      void set_revision( uint64_t revision ) {
+         if( _undo_stack.size() != 0 )
+            BOOST_THROW_EXCEPTION( std::logic_error("cannot set revision while there is an existing undo stack") );
+
+         if( revision > std::numeric_limits<int64_t>::max() )
+            BOOST_THROW_EXCEPTION( std::logic_error("revision to set is too high") );
+
+         if( revision < _revision )
+            BOOST_THROW_EXCEPTION( std::logic_error("revision cannot decrease") );
+
+         _revision = static_cast<int64_t>(revision);
+      }
+
+      std::pair<int64_t, int64_t> undo_stack_revision_range() const {
+         return { _revision - _undo_stack.size(), _revision };
+      }
+
+      /**
+       * Discards all undo history prior to revision
+       */
+      void commit( int64_t revision ) noexcept {
+         revision = std::min(revision, _revision);
+         if (revision == _revision) {
+            dispose(_old_values.before_begin(), _removed_values.before_begin());
+            _undo_stack.clear();
+         } else {
+            auto iter = _undo_stack.begin() + (_undo_stack.size() - (_revision - revision));
+            dispose(_old_values.iterator_to(*iter->old_values_end), _removed_values.iterator_to(*iter->removed_values_end));
+            _undo_stack.erase(_undo_stack.begin(), iter);
+         }
+      }
+
+      const undo_index& indices() const { return *this; }
+      template<typename Tag>
+      const auto& get() const { return std::get<index_of_tag<Tag, Keys...>::value>(_indices); }
+
+      template<int N>
+      const auto& get() const { return std::get<N>(_indices); }
+
+      std::size_t size() const {
+         return std::get<0>(_indices).size();
+      }
+
+      bool empty() const {
+         return std::get<0>(_indices).empty();
+      }
+
+      template<typename Tag, typename Iter>
+      auto project(Iter iter) const {
+         return get<Tag>().iterator_to(*iter);
+      }
+
+      const auto& stack() const { return _undo_stack; }
+
+      struct delta {
+         boost::iterator_range<typename index0_type::const_iterator> new_values;
+         boost::iterator_range<typename list_base<old_node, key0_type>::const_iterator> old_values;
+         boost::iterator_range<typename list_base<node, key0_type>::const_iterator> removed_values;
+      };
+
+      delta last_undo_session() const {
+        if(_undo_stack.empty())
+           return { { get<0>().end(), get<0>().end() },
+                    { _old_values.end(), _old_values.end() },
+                    { _removed_values.end(), _removed_values.end() } };
+         // FIXME: the problems with being lazy...
```

Comment:
Take care of this.

Reply:
The state_history_plugin needs this to be a const member, but it also needs to modify the internal state of the undo_index.  I'm not happy with the const_cast, but I don't have a better solution.

Reply:
If mutable is not an option, then I would prefer just a comment about why this is done. `FIXME` implies to me that you put this here as a marker to come back to. Before merging PR I would want the `FIXME` to be gone. Replaced with `// HACK: blah blah` for example would be better.